hypertrace-trace-enricher-api/src/main/java/org/hypertrace/traceenricher/trace/util/ApiTraceGraph.java

```diff
@@ -45,45 +44,39 @@ public ApiTraceGraph(StructuredTrace trace) {
     eventIdToIndexInTrace = buildEventIdToIndexInTrace(trace);
   }
 
-  public List<ApiNode> getNodeList() {
+  public List<ApiNode<Event>> getNodeList() {
     return nodeList;
   }
 
   public List<ApiNodeEventEdge> getApiNodeEventEdgeList() {
     return apiNodeEventEdgeList;
   }
 
-  public void setNodeList(List<ApiNode> nodeList) {
+  public void setNodeList(List<ApiNode<Event>> nodeList) {
     this.nodeList = nodeList;
   }
 
   // map of api node to index in api nodes list, since we want to build edges between api nodes
-  private Map<ApiNode, Integer> apiNodeToIndex;
+  private Map<ApiNode<Event>, Integer> apiNodeToIndex;
   //map of entry boundary event to api node. each api node has one entry boundary event
-  private Map<Event, ApiNode> entryBoundaryToApiNode;
+  private Map<Event, ApiNode<Event>> entryBoundaryToApiNode;
 
   /**
-   * The idea is to filter out the events in an API boundary.
-   * Filtered events can then be used to filter all the edges and entities from structured trace
-   * The filtered event, entities and edges will determine a small structured trace within an API boundary
+   * The idea is to filter out the events in an API boundary. The ApiNode will contain a head event for the API Trace,
+   * a list of events that fall within the API boundary, the Entry event or null if the head span is not an entry span
+   * and a list of exit events from the API.
    */
-  private List<ApiNode> buildApiNodes(StructuredTraceGraph graph, ByteBuffer traceId) {
-    List<ApiNode> apiNodes = new ArrayList<>();
+  private List<ApiNode<Event>> buildApiNodes(StructuredTraceGraph graph) {
+    List<ApiNode<Event>> apiNodes = new ArrayList<>();
     Set<ByteBuffer> processedEvents = new HashSet<>();
 
     for (Event event : trace.getEventList()) {
       if (EnrichedSpanUtils.isEntryApiBoundary(event)) {
         // get all the events in the api boundary
         Pair<List<Event>, List<Event>> pair = getEventsInApiBoundary(graph, event);
 
-        // create a new trace from the events
-        StructuredTrace newTrace = new StructuredTraceBuilder(
-            pair.getLeft(),
-            graph.getEntityMap(),
-            event.getCustomerId(),
-            traceId).buildStructuredTrace();
-
-        apiNodes.add(new ApiNode(newTrace, event, pair.getRight()));
+        // create new ApiNode from the events in the api boundary
+        apiNodes.add(new ApiNode<>(event, pair.getLeft(), event, pair.getRight()));
```

Comment:
Shouldn't the first argument be `trace.getEventList.get(0)`, because that's what was happening in ApiNode before for the headEvent?

Reply:
No. It should be event since that is the head of the API trace. trace.getEventList.get(0) is the full trace root event.